File: algorithms/heap-sort/heap-sort.rb
def left(i)
     return 2 * i + 1
end

def right(i)
     return 2 * i + 2
end

def max_heapify(alist, i, heap_size)
  l = left(i)
  r = right(i)

  if (l < heap_size && alist[l] > alist[i])
    largest = l
  else
    largest = i
  end 

  if (r < heap_size && alist[r] > alist[largest])
    largest = r
  end

  if (i != largest)
    temp = alist[i]
    alist[i] = alist[largest]
    alist[largest] = temp

    max_heapify(alist, largest, heap_size)
  end
end

def build_max_heap(alist)
  heap_size = alist.length

  (heap_size / 2 - 1).downto(0) do |i| 
    max_heapify(alist, i, heap_size)
  end

end

def heap_sort(alist)
  build_max_heap(alist)

  heap_size = alist.length
  while heap_size > 0
    heap_size -= 1

    temp = alist[0]
    alist[0] = alist[heap_size]
    alist[heap_size] = temp

    max_heapify(alist, 0, heap_size)
  end
end

alist = [1, 5, 3, 10, 35, 1, 4, 56]
heap_sort(alist)
print(alist)

File: algorithms/heap-sort/HeapSort.java
import java.util.Arrays;

public class HeapSort {
    public static int left(int i) {
        return 2 * i + 1;
    }

    public static int right(int i) {
        return 2 * i + 2;
    }

    public static void maxHeapify(int[] alist, int i, int heapSize) {
        int l = left(i);
        int r = right(i);

        int largest = i;
        if (l < heapSize && alist[l] > alist[i]) {
            largest = l;
        }
        if (r < heapSize && alist[r] > alist[largest]) {
            largest = r;
        }

        if (largest != i) {
            int temp = alist[i];
            alist[i] = alist[largest];
            alist[largest] = temp;

            maxHeapify(alist, largest, heapSize);
        }
        
    }

    public static void buildMaxHeap(int[] alist) {
        int heapSize = alist.length;
        for (int i = heapSize / 2 -1; i >= 0; i--) {
            maxHeapify(alist, i, heapSize);
            
        }
    }

    public static void sort(int[] alist) {
        int heapSize = alist.length;

        buildMaxHeap(alist);

        while (heapSize > 0) {
            heapSize --;

            int temp = alist[0];
            alist[0] = alist[heapSize];
            alist[heapSize] = temp;

            maxHeapify(alist, 0, heapSize);
        }
    }

    public static void main(String[] args) {
        int[] alist = {1, 2, 5, 89, 2, 22, 34, 11, 6};
        sort(alist);
        System.out.println(Arrays.toString(alist));
    }
}
